app/models/instance.rb

```diff
@@ -15,6 +15,9 @@ def find_tenant_by_host(host)
     end
   end
 
+  VALID_HOST_NAME_REGEX = /\A[a-z0-9]+([\-\.]{1}[a-z0-9]+)*\z/i
+  validates :host, format: { with: VALID_HOST_NAME_REGEX }, length: { maximum: 255 }
```

Comment:
So much discussion.

### Why gems
My main driver for using gems are that:
 1. They provide an abstraction barrier. You use the gem, not needing to know how it is implemented. This reduces the cognitive load on the programmer
 2. This follows the single responsibility principle. The idea is, your code should handle one level of abstraction, no more, no less. Gems take the logic that you would otherwise have to implement in your application and factors it out into reusable components. In this case, models are closely associated with the database, but validation logic has little to do with validating what is a valid domain name.
 3. Helps cohesion, as a specific instance of point 2.

### Using many gems
That is true. But remember that gems are supposed to be functionally cohesive and should do one thing. Gems are meant to be reusable across projects and are meant to be general. If the gem does not solve the problem that you wished the gem solves, the standard procedure is to fork the gem and submit a patch. Rails was built up as a framework in this manner.

If you think about this further, the 'hard to understand the code' reason then seems shaky. If a programmer finds that the code is hard to understand because of the various abstractions that have been used in a large-scale program, then I think he needs to learn how to manage complexity better. Having to deal with so many things at once would be _worse_ for maintainability.

For example, in this case: if I'm looking at the Instance model, and I either:
 - see a call saying it's validated as a hostname. I would skip it and say, okay, assume that works. If everything else fails, I check the gem (how often does this happen...?)
 - Otherwise, I see that regex. I have to figure out what the regex means, and figure out why the regex is used. Then I add a comment saying what the regex does. Then the regex gets changed. Then the regex and the comment doesn't sync up... I leave it alone. Then the next programmer comes. Rinse and repeat.
   - Try doing this everywhere in your application!

Once again, I think we are considering too much about lowering the barrier to entry for contributors. That is useful, and I've made such a decisions in the past; but history and experience shows that making things accessible brings absolutely no new contributors. If your have a complex project that solves problems, people will come. Perhaps that is true from a systems programming perspective (definitely true when I was writing kernel modules and other components in C++) but when you have a large scale application, this reason would prevent the application from doing useful things. _Complexity is inevitable, managing it through abstractions is how you ensure you can deal with it!_

### What to consider when using gems

 1. Using a little of the gem should not preclude the use of a gem. How much of Rails do you think we will be using? :smile: 
 2. How much time we would need to write the code out would probably be a concern. But remember that by doing this you are merely trading short-term speed for long-term maintenance. While this may be the agile way (don't write what you don't need), I think that for most instances, the barrier to using a gem is actually quite low, and you'll be spending about the same amount of time anyway.
 3. Future development is important.
 4. Maintenance is important. But sometimes a gem is just too good not to use. In the worst case, fork and _be the maintainer_. This makes you a better programmer and software engineer as well. You make decisions that typically you might expect a more senior programmer to make. You'll make mistakes. You'll learn. And then you won't. And then _you'll_ be the senior programmer.

Ultimately, whether to use a gem is a subjective opinion, and depends on the experiences of the people who have used others. Having used good and bad gems helps in making that decision. It's not necessarily something that can be done via consensus, unfortunately.